pytorch_lightning/loops/batch/manual.py

```diff
@@ -0,0 +1,92 @@
+    _check_training_step_output,
+    _process_training_step_output,
+)
+from pytorch_lightning.trainer.connectors.logger_connector.result import ResultCollection
+
+
+class ManualOptimization(Loop):
+    """A special loop implementing what is known in Lightning as Manual Optimization where the optimization happens
+    entirely in the :meth:`~pytorch_lightning.core.lightning.LightningModule.training_step` and therefore the user
+    is responsible for back-propagating gradients and making calls to the optimizers.
+
+    This loop is a trivial case because it performs only a single iteration (calling directly into the module's
+    :meth:`~pytorch_lightning.core.lightning.LightningModule.training_step`) and passing through the output(s).
+    """
+
+    def __init__(self) -> None:
+        super().__init__()
+        self._done: bool = False
+        self._hiddens: Optional[Any] = None
+        self._output: Optional[ResultCollection] = None
+
+    @property
+    def done(self) -> bool:
+        return self._done
+
+    def reset(self) -> None:
+        self._done = False
+
+    def advance(self, batch: Any, batch_idx: int, hiddens: Optional[Any] = None) -> None:  # type: ignore[override]
+        """Performs the training step for manual optimization.
+
+        Args:
+            batch: the current tbptt split of the current batch
+            batch_idx: the index of the current batch
+            hiddens: the model's hidden state of the previous iteration
+
+        Returns:
+            post-processed outputs from the training step, or ``None`` if training step returned nothing
+        """
+        assert self.trainer is not None
+        ligtning_module = self.trainer.lightning_module
+
+        with self.trainer.profiler.profile("model_forward"):
+
+            step_kwargs = _build_training_step_kwargs(
+                ligtning_module, self.trainer.optimizers, batch, batch_idx, opt_idx=None, hiddens=hiddens
+            )
+
+            # manually capture logged metrics
+            ligtning_module._current_fx_name = "training_step"
+            with self.trainer.profiler.profile("training_step"):
+                training_step_output = self.trainer.accelerator.training_step(step_kwargs)
+                self.trainer.accelerator.post_training_step()
+
+            del step_kwargs
+
+            training_step_output = self.trainer.call_hook("training_step_end", training_step_output)
+
+            _check_training_step_output(ligtning_module, training_step_output)
+
+            result_collection, hiddens = _process_training_step_output(self.trainer, training_step_output)
+
+        self._done = True
+        self._hiddens = hiddens
+        self._output = result_collection
+
+    def on_run_end(self) -> Optional[Tuple[ResultCollection, Optional[Any]]]:
+        hiddens = self._hiddens
+        output = self._output
+        self._hiddens, self._output = None, None  # free memory
+        if output is None:
+            return None
```

Comment:
Why not return the tuple directly?

Reply:
I think it's easier to check if the output of a function is None than to check it for each item in the tuple :)

Reply:
Should manual optimization allow returning `{loss: None, hiddens: something()}`?

Should anything returned by manual optimization be considered and `extra` and treated as such? What about the hiddens in that case? And the loss?

I think we should sit down and properly define these things because the integration of all these components is currently undefined and untested.